crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -193,37 +212,36 @@ impl SimpleFaucet {
         let number_of_coins = amounts.len();
         let total_amount: u64 = amounts.iter().sum();
 
-        let gas_coin = self
-            .prepare_gas_coin(total_amount, uuid)
-            .await
-            .tap_ok(|res| {
-                info!(recipient=?recipient, ?uuid, "Planning to use gas coin: {:?}", res);
-            })
-            .tap_err(|err| {
-                error!(
-                    ?uuid,
-                    "Failed to prepare gas coin with err: {:?}",
-                    err.to_string()
-                )
-            })
-            .map_err(|err| {
-                FaucetError::Internal(format!(
-                    "Failed to prepare gas coin with err: {:?}",
-                    err.to_string()
-                ))
-            })?;
-
-        let result = self
-            .execute_pay_sui_txn_with_retrials(
-                gas_coin,
-                self.active_address,
-                recipient,
-                amounts,
-                DEFAULT_GAS_BUDGET,
-                uuid,
-            )
-            .await;
+        let gas_coin_response = self.prepare_gas_coin(total_amount, uuid).await;
+
+        match gas_coin_response {
+            GasCoinResponse::ValidGasCoin(gas_coin) => {
+                let result = self
+                    .execute_pay_sui_txn_with_retrials(
+                        gas_coin,
+                        self.active_address,
+                        recipient,
+                        amounts,
+                        DEFAULT_GAS_BUDGET,
+                        uuid,
+                    )
+                    .await;
+                self.recycle_gas_coin(gas_coin, uuid).await;
+                self.check_and_map_transfer_gas_result(result, number_of_coins, &recipient)
+                    .await
+            }
+            GasCoinResponse::UnkownGasCoin(gas_coin) => {
```

Comment:
this is the fix, upon `UnkownGasCoin` due to fullnode reading failure, one coin has been consumed and we should recycle it.